### ISSUE TITLE
APOLLO optimizer

### ISSUE THREAD

**Issue by JaheimLee** [LoganBooker/prodigy-plus-schedule-free#12]
Could you integrate [APOLLO](https://github.com/zhuhanqing/APOLLO) to further reduce the memory usage? Maybe it can fully replace Adafactor?

**Comment by LoganBooker** [comment]
Hi @JaheimLee, thanks for the heads-up. I've had a look at APOLLO -- I don't think there's an elegant way to implement it Prodigy+SF. The main issue is how SF does momentum. While we can store a low-rank version of `z`, I don't see a good way of reshaping it for the update of `y`. The second moment is possible, but I'm not sure if the memory gains are worth the additional complexity. I will keep it in mind however.

**Comment by LoganBooker** [comment]
Possibly -- it looks like Prodigy will be getting some significant updates in the next few weeks (going from this comment https://github.com/konstmish/prodigy/pull/28#issuecomment-2578304552). I might wait until those are implemented and do it as part of a larger update.

**Comment by LoganBooker** [comment]
I've put a very experimental version in a new branch: https://github.com/LoganBooker/prodigy-plus-schedule-free/tree/experimental_galore_z. Rather than APOLLO, this uses a modified version of GaLore (upon which APOLLO is based) to store SF's underlying momentum in a lower rank.

I'd start with `galore_rank_denom=2` -- this should reduce the size of `z` by 50%. I've only tested it with LoRAs, I have my doubts about how it'll go with finetuning. :P

**Comment by JaheimLee** [comment]
Do you think using [radam_schedulefree](https://github.com/facebookresearch/schedule_free/blob/main/schedulefree/radam_schedulefree.py) is a good idea?

**Comment by LoganBooker** [comment]
`safeguard_warmup` in Prodigy is to stop the external LR scheduler from interfering with the adaptive stepsize. As SF uses a constant schedule, it didn't make much sense to include it in this optimiser.

As for RAdam -- in my tests it significantly dampens learning for an extended period (as intended), which pretty much kills Prodigy's ability to adapt the stepsize. Though this is less relevant with `use_speed=True`.

I'll have to think on it some more -- again, I'm not convinced storing `z` as a low rank approximation is the best approach. It sort-of works in the case of LoRAs, where the initial weight is zero, but for finetuning I think it might be too destructive, as SF uses a copy of `p` (the initial weights) to initialise `z`.

**Comment by LoganBooker** [comment]
@JaheimLee I've added RAdam to the optimiser as of this commit: https://github.com/LoganBooker/prodigy-plus-schedule-free/commit/55048b2f7a3258fbe06e1525d3958e8315c5b545. As noted, it will cripple Prodigy's LR calculations while the warmup is in effect, but for long training runs, this is probably not an issue. FYI, the optimiser's internal updates by default use `safeguard_warmup` and have since the first version.

I'll revisit GaLore, APOLLO, etc. down the track -- there might be an opportunity to use it when FOCUS is enabled, again, will need to experiment.
